File: integrations/terminal-sdk/python/tapntally_terminal.py
"""TapNTally terminal reference client (Python 3.9+, standard library only).

    from tapntally_terminal import build_bill, sign_bill, ndef_payload, push_bill

    bill = build_bill(bill_id="RF-KOR-000123", merchant={...}, terminal={...}, items=[...])
    signed = sign_bill(bill, terminal_secret)
    push_bill(api_base, api_key, pairing_code, signed)

Canonical JSON matches the Node and server implementations: keys sorted
recursively, no whitespace, None values omitted, UTF-8 unescaped.
"""
from __future__ import annotations

import base64
import hashlib
import hmac
import json
import urllib.error
import urllib.request
from datetime import datetime, timezone
from typing import Any

MIME = "application/vnd.tapntally.bill+json"


def _strip_none(value: Any) -> Any:
    if isinstance(value, dict):
        return {k: _strip_none(v) for k, v in value.items() if v is not None}
    if isinstance(value, list):
        return [_strip_none(v) for v in value]
    return value


def canonical_json(bill: dict) -> str:
    body = {k: v for k, v in bill.items() if k != "sig"}
    return json.dumps(_strip_none(body), sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def sign_bill(bill: dict, terminal_secret: str) -> dict:
    digest = hmac.new(terminal_secret.encode("utf-8"), canonical_json(bill).encode("utf-8"), hashlib.sha256).digest()
    return {**bill, "sig": base64.b64encode(digest).decode("ascii")}


def verify_bill(bill: dict, terminal_secret: str) -> bool:
    if not bill.get("sig"):
        return False
    expected = hmac.new(terminal_secret.encode("utf-8"), canonical_json(bill).encode("utf-8"), hashlib.sha256).digest()
    return hmac.compare_digest(base64.b64decode(bill["sig"]), expected)


def build_bill(
    *,
    bill_id: str,
    merchant: dict,
    terminal: dict,
    items: list[dict],
    tax_rupees: float = 0,
    discount_rupees: float = 0,
    payment_method: str = "unknown",
    issued_at: datetime | None = None,
) -> dict:
    """Amounts in rupees in; integer paise out, with totals that always reconcile."""

    def paise(r: float) -> int:
        return int(round(float(r) * 100))

    lines = []
    for it in items:
        unit = paise(it["unit_rupees"])
        line = {"name": it["name"], "qty": it["qty"], "unitPaise": unit, "totalPaise": int(round(unit * it["qty"]))}
        if it.get("sku"):
            line["sku"] = it["sku"]
        if it.get("gst_rate") is not None:
            line["gstRate"] = it["gst_rate"]
        lines.append(line)
    subtotal = sum(l["totalPaise"] for l in lines)
    tax, discount = paise(tax_rupees), paise(discount_rupees)
    when = issued_at or datetime.now(timezone.utc)
    return {
        "v": 1,
        "billId": bill_id,
        "issuedAt": when.isoformat(),
        "currency": "INR",
        "merchant": merchant,
        "terminal": terminal,
        "items": lines,
        "subtotalPaise": subtotal,
        "taxPaise": tax,
        "discountPaise": discount,
        "totalPaise": subtotal + tax - discount,
        "paymentMethod": payment_method,
    }


def ndef_payload(bill: dict) -> bytes:
    """One MIME NDEF record: MB=1, ME=1, TNF=0x02."""
    type_bytes = MIME.encode("ascii")
    payload = json.dumps(bill, ensure_ascii=False).encode("utf-8")
    short = len(payload) < 256
    flags = 0x80 | 0x40 | (0x10 if short else 0) | 0x02
    length = bytes([len(payload)]) if short else len(payload).to_bytes(4, "big")
    return bytes([flags, len(type_bytes)]) + length + type_bytes + payload


def push_bill(api_base: str, api_key: str, pairing_code: str, bill: dict, timeout: float = 15) -> dict:
    req = urllib.request.Request(
        f"{api_base.rstrip('/')}/partner/bills",
        data=json.dumps({"pairingCode": pairing_code, "bill": bill}).encode("utf-8"),
        headers={"content-type": "application/json", "x-partner-key": api_key},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=timeout) as res:
            return json.loads(res.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8", "replace")
        raise RuntimeError(f"TapNTally rejected the bill ({e.code}): {body}") from None


if __name__ == "__main__":
    import argparse
    import sys

    p = argparse.ArgumentParser(description="TapNTally terminal client")
    p.add_argument("command", choices=["sign", "push"])
    p.add_argument("--bill", required=True)
    p.add_argument("--secret")
    p.add_argument("--api")
    p.add_argument("--key")
    p.add_argument("--code")
    a = p.parse_args()
    with open(a.bill, encoding="utf-8") as f:
        bill = json.load(f)
    if a.secret:
        bill = sign_bill(bill, a.secret)
    if a.command == "sign":
        json.dump(bill, sys.stdout, indent=2, ensure_ascii=False)
    else:
        print(json.dumps(push_bill(a.api, a.key, a.code, bill), indent=2))

File: integrations/terminal-sdk/node/tapntally-terminal.js
#!/usr/bin/env node
/**
 * TapNTally terminal reference client (Node 18+, zero dependencies).
 *
 *   const tnt = require('./tapntally-terminal');
 *   const bill = tnt.buildBill({ ... });                 // TBEF v1 object
 *   const signed = tnt.signBill(bill, terminalSecret);   // adds .sig
 *   const ndef = tnt.ndefPayload(signed);                // Buffer for NFC / HCE
 *   await tnt.pushBill({ apiBase, apiKey, pairingCode, bill: signed });
 *
 * CLI:
 *   node tapntally-terminal.js sign --secret <s> --bill bill.json
 *   node tapntally-terminal.js push --api <url> --key <tnt_…> --code <ABC123> --bill bill.json [--secret <s>]
 */
'use strict';
const crypto = require('node:crypto');
const fs = require('node:fs');

const MIME = 'application/vnd.tapntally.bill+json';

/** Recursively sort keys and drop undefined values so JSON.stringify is canonical. */
function canonicalize(value) {
  if (Array.isArray(value)) return value.map(canonicalize);
  if (value && typeof value === 'object') {
    const out = {};
    for (const key of Object.keys(value).sort()) {
      if (value[key] !== undefined) out[key] = canonicalize(value[key]);
    }
    return out;
  }
  return value;
}

function canonicalJson(bill) {
  const { sig, ...rest } = bill;
  return JSON.stringify(canonicalize(rest));
}

function signBill(bill, terminalSecret) {
  const sig = crypto.createHmac('sha256', terminalSecret).update(canonicalJson(bill), 'utf8').digest('base64');
  return { ...bill, sig };
}

function verifyBill(bill, terminalSecret) {
  if (!bill.sig) return false;
  const expected = crypto.createHmac('sha256', terminalSecret).update(canonicalJson(bill), 'utf8').digest();
  const given = Buffer.from(bill.sig, 'base64');
  return given.length === expected.length && crypto.timingSafeEqual(given, expected);
}

/**
 * Build a TBEF v1 bill from plain inputs. Amounts are rupees here and
 * converted to integer paise; totals are computed so they always reconcile.
 */
function buildBill({ billId, merchant, terminal, items, taxRupees = 0, discountRupees = 0, paymentMethod = 'unknown', issuedAt = new Date() }) {
  const toPaise = (r) => Math.round(Number(r) * 100);
  const lines = items.map((it) => {
    const unitPaise = toPaise(it.unitRupees);
    const totalPaise = Math.round(unitPaise * it.qty);
    return { name: it.name, qty: it.qty, unitPaise, totalPaise, ...(it.sku ? { sku: it.sku } : {}), ...(it.gstRate !== undefined ? { gstRate: it.gstRate } : {}) };
  });
  const subtotalPaise = lines.reduce((s, l) => s + l.totalPaise, 0);
  const taxPaise = toPaise(taxRupees);
  const discountPaise = toPaise(discountRupees);
  return {
    v: 1,
    billId,
    issuedAt: issuedAt.toISOString(),
    currency: 'INR',
    merchant,
    terminal,
    items: lines,
    subtotalPaise,
    taxPaise,
    discountPaise,
    totalPaise: subtotalPaise + taxPaise - discountPaise,
    paymentMethod,
  };
}

/** Raw NDEF message (one MIME record) ready to write to a tag or return from an HCE READ. */
function ndefPayload(bill) {
  const type = Buffer.from(MIME, 'ascii');
  const payload = Buffer.from(JSON.stringify(bill), 'utf8');
  const short = payload.length < 256;
  // Flags: MB=1, ME=1, SR=short, TNF=0x02 (MIME media)
  const flags = 0x80 | 0x40 | (short ? 0x10 : 0) | 0x02;
  const header = [flags, type.length];
  const lenBytes = short ? Buffer.from([payload.length]) : Buffer.from([(payload.length >>> 24) & 255, (payload.length >>> 16) & 255, (payload.length >>> 8) & 255, payload.length & 255]);
  return Buffer.concat([Buffer.from(header), lenBytes, type, payload]);
}

/** Server-to-server delivery with a customer pairing code. Resolves to { transactionId, created, categorisedAs }. */
async function pushBill({ apiBase, apiKey, pairingCode, bill, fetchImpl = globalThis.fetch }) {
  const res = await fetchImpl(`${apiBase.replace(/\/$/, '')}/partner/bills`, {
    method: 'POST',
    headers: { 'content-type': 'application/json', 'x-partner-key': apiKey },
    body: JSON.stringify({ pairingCode, bill }),
  });
  const body = await res.json().catch(() => ({}));
  if (!res.ok) {
    const err = new Error(`TapNTally rejected the bill: ${body.code ?? res.status} ${body.message ?? ''}`.trim());
    err.status = res.status;
    err.body = body;
    throw err;
  }
  return body;
}

module.exports = { MIME, canonicalJson, signBill, verifyBill, buildBill, ndefPayload, pushBill };

// ---- CLI -------------------------------------------------------------------
if (require.main === module) {
  const [cmd, ...rest] = process.argv.slice(2);
  const args = Object.fromEntries(rest.map((a, i, arr) => (a.startsWith('--') ? [a.slice(2), arr[i + 1]] : null)).filter(Boolean));
  const readBill = () => JSON.parse(fs.readFileSync(args.bill, 'utf8'));
  (async () => {
    if (cmd === 'sign') {
      process.stdout.write(JSON.stringify(signBill(readBill(), args.secret), null, 2) + '\n');
    } else if (cmd === 'push') {
      let bill = readBill();
      if (args.secret) bill = signBill(bill, args.secret);
      const out = await pushBill({ apiBase: args.api, apiKey: args.key, pairingCode: args.code, bill });
      console.log(JSON.stringify(out, null, 2));
    } else if (cmd === 'ndef') {
      process.stdout.write(ndefPayload(readBill()));
    } else {
      console.error('usage: tapntally-terminal.js <sign|push|ndef> --bill file [--secret s] [--api url --key k --code c]');
      process.exit(2);
    }
  })().catch((e) => {
    console.error(e.message);
    process.exit(1);
  });
}
